File: python/Q338_counting_bits.py
#
# @lc app=leetcode id=338 lang=python3
#
# [338] Counting Bits
#

# @lc code=start
class Solution:
    def countBits(self, n: int) -> List[int]:
        result: list[int] = [0, 1]

        while len(result) < n + 1:
            result += [n + 1 for n in result]

        return result[: n + 1]


# @lc code=end

File: rust/src/leetcode/Q338_counting_bits.rs
/*
 * @lc app=leetcode id=338 lang=rust
 *
 * [338] Counting Bits
 */

// @lc code=start
impl Solution {
    pub fn count_bits(n: i32) -> Vec<i32> {
        let mut result = vec![0; (n + 1) as usize];
        
        for i in 1..=n {
            result[i as usize] = result[(i >> 1) as usize] + (i & 1);
        }
        
        result
    }
}
// @lc code=end

